wgpu-core/src/command/ray_tracing.rs

```diff
@@ -0,0 +1,1225 @@
+use crate::{
+    device::queue::TempResource,
+    global::Global,
+    id::CommandEncoderId,
+    init_tracker::MemoryInitKind,
+    lock::RwLockReadGuard,
+    ray_tracing::{
+        tlas_instance_into_bytes, BlasAction, BlasBuildEntry, BlasGeometries,
+        BuildAccelerationStructureError, TlasAction, TlasBuildEntry, TlasPackage,
+        ValidateBlasActionsError, ValidateTlasActionsError,
+    },
+    resource::{Blas, Tlas},
+    FastHashSet,
+};
+
+use wgt::{math::align_to, BufferAddress, BufferUsages, Features};
+
+use super::{BakedCommands, CommandBufferMutable};
+use crate::ray_tracing::BlasTriangleGeometry;
+use crate::resource::{
+    AccelerationStructure, Buffer, Fallible, Labeled, ScratchBuffer, StagingBuffer, Trackable,
+};
+use crate::snatch::SnatchGuard;
+use crate::storage::Storage;
+use crate::track::PendingTransition;
+use hal::BufferUses;
+use std::ops::Deref;
+use std::sync::Arc;
+use std::{cmp::max, num::NonZeroU64, ops::Range};
+
+type BufferStorage<'a> = Vec<(
+    Arc<Buffer>,
+    Option<PendingTransition<BufferUses>>,
+    Option<(Arc<Buffer>, Option<PendingTransition<BufferUses>>)>,
+    Option<(Arc<Buffer>, Option<PendingTransition<BufferUses>>)>,
+    BlasTriangleGeometry<'a>,
+    Option<Arc<Blas>>,
+)>;
+
+type BlasStorage<'a> = Vec<(
+    Arc<Blas>,
+    hal::AccelerationStructureEntries<'a, dyn hal::DynBuffer>,
+    u64,
+)>;
+
+// This should be queried from the device, maybe the the hal api should pre aline it, since I am unsure how else we can idiomatically get this value.
+const SCRATCH_BUFFER_ALIGNMENT: u32 = 256;
+
+impl Global {
+    pub fn command_encoder_build_acceleration_structures_unsafe_tlas<'a>(
+        &self,
+        command_encoder_id: CommandEncoderId,
+        blas_iter: impl Iterator<Item = BlasBuildEntry<'a>>,
+        tlas_iter: impl Iterator<Item = TlasBuildEntry>,
+    ) -> Result<(), BuildAccelerationStructureError> {
+        profiling::scope!("CommandEncoder::build_acceleration_structures_unsafe_tlas");
+
+        let hub = &self.hub;
+
+        let cmd_buf = hub
+            .command_buffers
+            .get(command_encoder_id.into_command_buffer_id());
+
+        let buffer_guard = hub.buffers.read();
+        let blas_guard = hub.blas_s.read();
+        let tlas_guard = hub.tlas_s.read();
+
+        let device = &cmd_buf.device;
+
+        if !device
+            .features
+            .contains(Features::RAY_TRACING_ACCELERATION_STRUCTURE)
+        {
+            return Err(BuildAccelerationStructureError::MissingFeature);
+        }
+
+        let build_command_index = NonZeroU64::new(
+            device
+                .last_acceleration_structure_build_command_index
+                .fetch_add(1, std::sync::atomic::Ordering::Relaxed)
+                + 1,
+        )
+        .unwrap();
+
+        #[cfg(feature = "trace")]
```

Comment:
this is complicated, due to iterators taking things by value the trace BLAS and TLAS currently need to be recreated, which cannot be done this way, if you prefer this way, this could go like queue submit with the last level taking a slice.

Reply:
Fair enough.